src/shaders/globe_raster.vertex.glsl

```diff
@@ -1,33 +1,90 @@
+#define HALF_PI PI / 2.0
+#define QUARTER_PI PI / 4.0
+#define RAD_TO_DEG 180.0 / PI
+#define DEG_TO_RAD PI / 180.0
+#define GLOBE_RADIUS EXTENT / PI / 2.0
+
 uniform mat4 u_proj_matrix;
 uniform mat4 u_globe_matrix;
 uniform mat4 u_merc_matrix;
 uniform float u_zoom_transition;
 uniform vec2 u_merc_center;
+uniform mat3 u_grid_matrix;
 
+#ifdef GLOBE_POLES
 attribute vec3 a_globe_pos;
 attribute vec2 a_merc_pos;
 attribute vec2 a_uv;
+#else
+attribute vec2 a_pos;
+#endif
 
 varying vec2 v_pos0;
 
 const float wireframeOffset = 1e3;
 
+float mercatorXfromLng(float lng) {
+    return (180.0 + lng) / 360.0;
+}
+
+float mercatorYfromLat(float lat) {
+    return (180.0 - (RAD_TO_DEG* log(tan(QUARTER_PI + lat / 2.0 * DEG_TO_RAD)))) / 360.0;
+}
+
+vec3 latLngToECEF(vec2 latLng) {
+    latLng = DEG_TO_RAD * latLng;
+    
+    float cosLat = cos(latLng[0]);
+    float sinLat = sin(latLng[0]);
+    float cosLng = cos(latLng[1]);
+    float sinLng = sin(latLng[1]);
+
+    // Convert lat & lng to spherical representation. Use zoom=0 as a reference
+    float sx = cosLat * sinLng * GLOBE_RADIUS;
+    float sy = -sinLat * GLOBE_RADIUS;
+    float sz = cosLat * cosLng * GLOBE_RADIUS;
+
+    return vec3(sx, sy, sz);
+}
+
 void main() {
-    v_pos0 = a_uv;
+#ifdef GLOBE_POLES
+    vec3 globe_pos = a_globe_pos;
+    vec2 merc_pos = a_merc_pos;
+    vec2 uv = a_uv;
+#else
+    float tiles = u_grid_matrix[0][2];
+    float idy = u_grid_matrix[1][2];
+    float S = u_grid_matrix[2][2];
```

Comment:
Why encoding these in the `mat3` and not in their own uniforms? It should be possible to consolidate these as their own `vec3`. Otherwise, I find it a bit confusing as part of the `mat3` as they aren't affecting the transform at https://github.com/mapbox/mapbox-gl-js/pull/11560/files#diff-184e79e72d9cb77dd6f755ab3fc6f022fde3fee19cd63971da8672ef0bd2a834R60.

Reply:
Actually I did put them in their own respective uniforms before but I refactored it to the current version since the 3x3 matrix had some spare space that could be used for this purpose.